README.md
## cystiSim

[![Travis-CI Build Status](https://travis-ci.org/brechtdv/cystiSim.svg?branch=master)](https://travis-ci.org/brechtdv/cystiSim)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/cystiSim)](http://cran.r-project.org/package=cystiSim)
[![Cranlogs Downloads](http://cranlogs.r-pkg.org/badges/grand-total/cystiSim)](http://cran.r-project.org/package=cystiSim)

_Agent-based model for Taenia solium transmission and control_

#### Download and install

```r
devtools::install_github("brechtdv/cystiSim")
library("cystiSim")
```

#### Getting started

Start exploring `cystiSim` through the documentation and examples on the [Wiki page](https://github.com/brechtdv/cystiSim/wiki).

#### Acknowledgments

[`cystiSim`](https://github.com/brechtdv/cystiSim) was developed by [@UCBraae](https://twitter.com/ucbraae) and [@brechtdv](https://github.com/brechtdv), within the framework of [CYSTINET](http://www.cystinet.org/), the European Network on taeniosis/cysticercosis, COST ACTION TD1302.


NEWS
cystiSim 0.1.2
==============

* set `stringsAsFactors=FALSE` in data.frames

cystiSim 0.1.1
==============

* fix bug in `update` method (no slaughter == no births)

cystiSim 0.1.0
==============

* first commit to CRAN

cystiSim 0.0.6
==============

* rename intervention functions

* update help files

* add argument `annotate` to `plot.cystiSim`

cystiSim 0.0.5
==============

* add `elim` method for `cystiSim` objects

* fix bug in `exp_decay`

cystiSim 0.0.4
==============

* fix bug in definition of `slaughter.args`

* fix subsetting bug in `do_oxf_vac`

cystiSim 0.0.3
==============

* various CRAN error/warning/note fixes

* change default `min.age` in `do_oxf` to 1

* introduce `do_oxf_vac` function

cystiSim 0.0.2
==============

* introduce `magrittr` style model definitions

* introduce functionalities and methods for multiple `cystiSim` simulations

* introduce functionalities for fitting models to data

* update model to make `slaughter` function generic


cystiSim 0.0.1
==============

* the first version

R/fit.R
### CYSTISIM: AGENT BASED TAENIA SOLIUM TRANSMISSION MODEL

## -------------------------------------------------------------------------#
## FIT MODEL PARAMETERS ----------------------------------------------------#

fit <-
function(n.sim, n.update, target, limit,
         man, pig, ph2m, pl2m, m2p, e2p, age.coef = c(0, 0),
         slaughter = slaughter_nbinom,
         slaughter.args = list(min = 6,  max = 36, size = 0.70, mu = 80)) {
  ## evaluate 'target'
  if (!is.list(target)) stop("'target' must be a list")
  sapply(names(target), match.arg, c("ht", "pc", "pi"))

  ## predefine 
  out <- matrix(ncol = 16, nrow = 0)
  colnames(out) <- c("ph2m", "pl2m", "m2p", "e2p", 
                     "PC", "P(H)", "P(L)", "PCI", "PI",
                     "HT", "HT(inf)", "HT(ch)", "HT(ad)", "EN", "HT(tot)",
                     "DEV")

  ## initiate counter
  i <- 0

  ## initiate progress bar
  pb <- txtProgressBar(max = n.sim, style = 3)

  ## run simulations
  while (i < n.sim) {
    ph2m_sim <- runif(1, ph2m[1], ph2m[2])
    pl2m_sim <- runif(1, pl2m[1], pl2m[2])
    m2p_sim  <- runif(1, m2p[1], m2p[2])
    e2p_sim  <- runif(1, e2p[1], e2p[2])

    mod <-
      initiate(man, pig,
               ph2m_sim, pl2m_sim, m2p_sim, e2p_sim, age.coef,
               slaughter, slaughter.args) %>%
      update(n.update, verbose = FALSE)

    ht <- prevalence(mod$man$taenia_immature + mod$man$taenia)
    pc <- prevalence(mod$pig$cysti)
    pi <- sum(mod$pig$intensity == "H") / sum(mod$pig$intensity != "0")

    ht_diff <- ifelse("ht" %in% names(target), ht-target$ht, 0)
    pc_diff <- ifelse("pc" %in% names(target), pc-target$pc, 0)
    pi_diff <- ifelse("pi" %in% names(target), pi-target$pi, 0)
    if (is.nan(pi_diff)) pi_diff <- 0

    dev <- ht_diff^2 + pc_diff^2 + pi_diff^2

    if (dev < limit) {
      i <- i + 1
      out <- rbind(out,
                   c(ph2m_sim, pl2m_sim, m2p_sim, e2p_sim,
                   c(tail(mod$out, 1)), ht, dev))
      setTxtProgressBar(pb, i)
    }
  }
  
  ## close progress bar
  close(pb)

  cat("\n")

  return(out)
}


R/sim.R
### CYSTISIM: AGENT BASED TAENIA SOLIUM TRANSMISSION MODEL

## -------------------------------------------------------------------------#
## MAIN SIMULATION FUNCTION ------------------------------------------------#

cystiSim <-
function(n = 100, mod, main = NULL) {
  ## predefine list of simulations
  out <- vector("list", n)

  ## perform simulations
  for (i in seq(n)) {
    print(i)
    out[[i]] <- eval(substitute(mod))$out[, c(1, 5, 6)]
  }

  ## create 'cystiSims' object
  sim <-
    list(out = out,
         mod = substitute(mod),
         main = main,
         par = list(ph2m = ph2m,
                    pl2m = pl2m,
                    m2p = m2p,
                    e2p = e2p,
                    eff_mda = eff_mda, cov_mda = cov_mda,
                    eff_oxf = eff_oxf, cov_oxf = cov_oxf,
                    eff_vac = eff_vac, cov_vac = cov_vac))

  ## add S3 class 'cystiSims'
  class(sim) <- "cystiSim"

  ## return 'cystiSims' object
  return(sim)
}


## -------------------------------------------------------------------------#
## cystiSims METHODS -------------------------------------------------------#

print.cystiSim <-
function(sim) {
  mod <- strsplit(as.character(sim$mod)[2], "%>%")[[1]]
  mod <- gsub("\n", "", mod)
  mod <- sub("^ +", "", mod)
  mod <- sub(" +$", "", mod)
  for (i in seq_along(mod)) {
    if (i > 1) cat ("  ")
    cat(mod[i])
    if (i < length(mod)) cat(" %>%\n")
  }
  cat("\n")
}

summary.cystiSim <-
function(sim, round = 3) {
  sim_mx <- sapply(sim$out, as.matrix)
  m <- nrow(sim_mx) / ncol(sim$out[[1]])

  out <-
  rbind(
    c(mean(sim_mx[m, ] == 0), quantile(sim_mx[m, ], 0.95)),      # PC
    c(mean(sim_mx[3*m, ] == 0), quantile(sim_mx[3*m, ], 0.95)))  # HT

  rownames(out) <- c("PC", "HT")
  colnames(out) <- c("Pr(elim)", "95%")

  return(round(out, round))
}

plot.cystiSim <-
function(sim) {
  lab <- factor(c("PC", "PR", "HT"), c("PC", "PR", "HT"))
  grp <- factor(c("pig", "pig", "human"), c("pig", "human"))
  col <- c(1, 3, 2)

  sim_mx <- sapply(sim$out, as.matrix)  # ncol=it; nrow=par*months

  df <-
  data.frame(mean = rowMeans(sim_mx),
             lwr  = apply(sim_mx, 1, quantile, 0.025),
             upr  = apply(sim_mx, 1, quantile, 0.975),
             m = rep(seq(nrow(sim_mx) / ncol(sim$out[[1]])),
                     ncol(sim$out[[1]])),
             par = rep(lab, each = nrow(sim_mx) / ncol(sim$out[[1]])),
             grp = rep(grp, each = nrow(sim_mx) / ncol(sim$out[[1]])))

  stats <- summary(sim)

  x_txt <- nrow(sim_mx) / ncol(sim$out[[1]])
  y_txt <- tapply(apply(sim_mx, 1, max), df$grp, max)

  df_txt <-
    data.frame(lab = paste0(ncol(sim_mx), " simulations",
                            "\nPr(elim)=", stats[, 1]),
               grp = levels(df$grp),
               x = x_txt,
               y = y_txt)

  ggplot(df, aes_string(x = "m", y = "mean")) +
    geom_ribbon(aes_string(ymin = "lwr", ymax = "upr", fill = "par"),
                alpha = .25) +
    geom_line(aes_string(y = "mean", col = "par"), size = 1) +
    geom_line(aes_string(y = "lwr", col = "par")) +
    geom_line(aes_string(y = "upr", col = "par")) +
    scale_fill_manual(values = col) +
    scale_colour_manual(values = col) +
    scale_x_continuous("month") +
    scale_y_continuous("prevalence") +
    facet_grid(grp~., scales = "free") +
    theme(legend.position = "none") +
    theme_bw() +
    ggtitle(sim$main) +
    geom_text(data = df_txt,
              aes_string(label = "lab", x = "x", y = "y"),
              size = 4, hjust = 1, vjust = 1)
}

report <-
function(sim, ...) {
  UseMethod("report")
}

report.cystiSim <-
function(sim, name = "cystiSim") {
  ## temporary 'mod' file
  ## write model
  tmp <- tempfile(fileext = ".Rnw")
  sink(tmp)
  cat("<<eval=FALSE>>=\n")  
  print(sim)
  cat("@\n")
  sink()

  ## knit PDF
  knit2pdf(system.file("cystiSim.Rnw", package = "cystiSim"))

  ## rename PDF
  file.rename("cystiSim.pdf",
              paste0(name, "_", today(), ".pdf"))

  ## remove helper files
  unlink(tmp)
  file.remove(paste0("cystiSim.",
                     c("tex", "log", "aux")))
  unlink("figure", recursive = TRUE)

  ## generate PNG
  png(paste0(name, "_", today(), ".png"),
      width = 10, height = 5, units = "in", res = 300)
  print(plot(sim))
  graphics.off()
}


## -------------------------------------------------------------------------#
## HELPER FUNCTIONS --------------------------------------------------------#

today <-
function() {
  return(format(Sys.time(), "%Y%m%d"))
}

R/model.R
### CYSTISIM: AGENT BASED TAENIA SOLIUM TRANSMISSION MODEL

## -------------------------------------------------------------------------#
## HELPER FUNCTIONS --------------------------------------------------------#

## expit = inverse logit
expit <-
function(x) {
  exp(x)/(1 + exp(x))
}

## reporter function: prevalence
prevalence <-
function(z) {
  sum(z == 1, na.rm = TRUE) / length(z[!is.na(z)])
}

## reporter function: census
census <-
function(x) {
    ## porcine cysticercosis (mature) prevalence
  c(prevalence(x$pig$cysti),

    ## proportion high infection intensity
    prevalence(x$pig$intensity[x$pig$cysti == 1] == "H"),

    ## proportion low infection intensity
    prevalence(x$pig$intensity[x$pig$cysti == 1] == "L"),

    ## porcine cysticercosis (immature) prevalence
    prevalence(x$pig$cysti_immature),

    ## proportion immune pigs
    prevalence(x$pig$immunity > 0),

    ## taeniosis (mature) prevalence
    prevalence(x$man$taenia),

    ## taeniosis (mature) prevalence, infants
    prevalence(x$man$taenia[x$man$age <= 3*12]),

    ## taeniosis (mature) prevalence, children
    prevalence(x$man$taenia[x$man$age > 3*12 & x$man$age < 16*12]),

    ## taeniosis (mature) prevalence, adults
    prevalence(x$man$taenia[x$man$age >= 16*12]),

    ## environmental contamination prevalence
    prevalence(x$man$environment))
}

## exponential decay of environmental contamination
exp_decay <-
function(time) {
  p_decay <- rep(0.2682422, length(time))
  p_decay[time >= 9] <- 1     # 100% decay at month 9
  return(p_decay)
}


## -------------------------------------------------------------------------#
## TRANSMISSION FUNCTIONS --------------------------------------------------#

infect_man <-
function(man, pig, ph2m, pl2m, age.coef) {
  ## calculate number of Pigs Infected / High intensity
  PIH <- with(pig, sum(cysti == 1 & slaughtered == 1 & intensity == "H"))

  ## calculate number of Pigs Infected / Low intensity
  PIL <- with(pig, sum(cysti == 1 & slaughtered == 1 & intensity == "L"))

  ## identify Susceptible Humans
  HS <- with(man,
             taenia == 0 & taenia_immature == 0 &  # without tapeworm
             age >= 24)                            # no infants < 2yo

  ## randomly create new Human Tapeworm carriers
  ## probability of HT through any pig
  ## == 1 - probability of no infection through any pig
  p2m <- 1 - (1 - (1 - (1 - ph2m) ^ PIH)) *
             (1 - (1 - (1 - pl2m) ^ PIL))

  ## 'p2m' is the average probability of infection
  ## disaggregate 'p2m' according to age

  ## maximum age in population
  age_max <- max(man$age[HS] / 12)

  ## age-specific odds of infection defined from logistic regression
  odds <- expit(age.coef[1] + seq(0, age_max) * age.coef[2])

  ## tabulate age distribution of current dataframe
  age_dist <- table(cut(man$age[HS] / 12, seq(0, age_max + 1), right = FALSE))

  ## define ratio of odds and age-weighted average
  ratio <- odds / weighted.mean(odds, age_dist)

  ## define age-specific probability of HT
  p2m_age <- p2m * ratio[1 + (man$age[HS] / 12)]

  ## now randomly create new Human Tapeworm carriers
  ## based on age-specific transmission probability
  new_HT <- rbinom(sum(HS), 1, p2m_age)

  ## update 'taenia_immature' identifier
  man$taenia_immature[HS] <- new_HT

  ## update 'time_since_infection' identifier (redundant)
  man$time_since_infection[HS][new_HT == 1] <- 0

  ## return updated dataframe
  return(man)
}

infect_pig <-
function(pig, man, m2p, e2p) {

  ## DIRECT TRANSMISSION

  ## calculate number of Human Tapeworm carriers
  HT <- with(man, sum(taenia))

  ## identify Susceptible Pigs
  PS <- with(pig,
             cysti == 0 & cysti_immature == 0 &     # no cysts
             immunity == 0 &                        # no immunity
             slaughtered == 0)                      # not slaughtered

  ## randomly create new PC High intensity cases (from Susceptible cases)
  new_PCS2H <- rbinom(sum(PS), 1, 1 - (1 - m2p) ^ HT)

  ## identify Low intensity Pigs
  PSL <- with(pig,
              (cysti == 1 | cysti_immature == 1) &  # cysts
              immunity == 0 &                       # no immunity (redundant)
              slaughtered == 0 &                    # not slaughtered
              intensity == "L")                     # low intensity

  ## randomly create new PC High intensity cases (from Low intensity cases)
  new_PCL2H <- rbinom(sum(PSL), 1, 1 - (1 - m2p) ^ HT)


  ## INDIRECT TRANSMISSION

  ## calculate number of contaminated ENvironments
  EN <- with(man, sum(environment))

  ## randomly create new PC Low intensity cases (from Susceptible cases)
  new_PCS2L <- rbinom(sum(PS), 1, 1 - (1 - e2p) ^ EN)


  ## UPDATE IDENTIFIERS

  ## susceptible pigs becoming infected through HT or EN
  new_PC <- 1 - ((1 - new_PCS2L) * (1 - new_PCS2H))

  ## update 'cysti_immature' identifier
  pig$cysti_immature[PS] <- new_PC

  ## update 'intensity' identifier
  pig$intensity[PS][new_PCS2L == 1] <- "L"   # new Low intensity case
  pig$intensity[PS][new_PCS2H == 1] <- "H"   # new High intensity case
  pig$intensity[PSL][new_PCL2H == 1] <- "H"  # update to High from Low

  ## update 'time_since_infection' identifier (redundant)
  pig$time_since_infection[PS][new_PC == 1] <- 0

  ## return updated dataframe
  return(pig)
}


## -------------------------------------------------------------------------#
## INITIATE MODEL ----------------------------------------------------------#

initiate <-
function(man, pig,
         ph2m, pl2m, m2p, e2p, age.coef = c(0, 0),
         slaughter = slaughter_nbinom,
         slaughter.args = list(min = 6, max = 36, size = 0.70, mu = 80)) {
  ## create 'cystiRun' object
  x <-
  list(man = man,
       pig = pig,
       out = NULL,
       par = list(age_coef = age.coef,
                  ph2m = ph2m,
                  pl2m = pl2m,
                  m2p = m2p,
                  e2p = e2p),
       slaughter = list(f = slaughter,
                        args = slaughter.args))

  ## take baseline population census
  x$out <- matrix(census(x), ncol = 10)

  ## add S3 class 'cystiRun'
  class(x) <- "cystiRun"

  ## return 'cystiRun' object
  return(x)
}


## -------------------------------------------------------------------------#
## UPDATE MODEL ------------------------------------------------------------#

update.cystiRun <-
function(object, n = 1200, verbose = TRUE, ...) {

## MODEL
x <- object
out <- matrix(nrow = n, ncol = 10)

if (verbose) pb <- txtProgressBar(max = n, style = 3)

for (i in seq(n)) {

  ## STEP 01: maturation of tapeworms present in previous cycle
  man_infected <- x$man$taenia == 1 | x$man$taenia_immature == 1
  x$man$time_since_infection[man_infected] <-
    x$man$time_since_infection[man_infected] + 1

  ## STEP 02: immature tapeworms become infectious after 2 months
  ##          and starts contaminating environment
  man_infectious <- x$man$time_since_infection == 2
  x$man$taenia[man_infectious] <- 1             ## mature HT
  x$man$taenia_immature[man_infectious] <- 0    ## no immature HT
  x$man$environment[man_infectious] <- 1        ## envir contam

  ## STEP 03: tapeworms die after 13 months and human becomes susceptible
  tw_dies <- x$man$time_since_infection == 13
  x$man$taenia[tw_dies] <- 0                     # tw is gone
  x$man$time_since_infection[tw_dies] <- 0       # reset time since inf
  x$man$time_since_contamination[tw_dies] <- 1   # envir contamination!

  ## STEP 04: exponential decay of environmental contamination
  ## .. if no HT & time.since.contam > 0, exp decay
  ## .. if no HT & time.since.contam > 0 & no decay, time.since.contam++
  is_decay <- x$man$time_since_contamination > 0
  p_decay  <- exp_decay(x$man$time_since_contamination[is_decay])
  decay    <- rbinom(sum(is_decay), 1, p_decay) == 1
  x$man$time_since_contamination[is_decay][decay] <- 0
  x$man$environment[is_decay][decay] <- 0  ### SET ENVIRONMENT TO 0
  x$man$time_since_contamination[is_decay][!decay] <-
    x$man$time_since_contamination[is_decay][!decay] + 1

  ## STEP 05: new HT, based on slaughtered infectious pigs of previous cycle
  man_new <- infect_man(x$man, x$pig,
                        x$par$ph2m, x$par$pl2m, x$par$age_coef)

  ## STEP 06: remove slaughtered pigs of previous cycle
  x$pig <- subset(x$pig, x$pig$slaughtered == 0)

  ## STEP 07: maturation of cysts present in previous cycle
  is_infected <- x$pig$cysti_immature == 1
  x$pig$time_since_infection[is_infected] <-
    x$pig$time_since_infection[is_infected] + 1

  ## STEP 08: immature cysts become infectious after 3 months
  is_infectious <- x$pig$time_since_infection == 3
  x$pig$cysti[is_infectious] <- 1
  x$pig$cysti_immature[is_infectious] <- 0
  x$pig$time_since_infection[is_infectious] <- 0

  ## STEP 09: new PI, based on HT & ENVIR of previous cycle
  pig_new <- infect_pig(x$pig, x$man, x$par$m2p, x$par$e2p)

  ## STEP 10: slaughter pigs of previous cycle
  is_slaughtered <-
    do.call(x$slaughter$f, c(list(age = pig_new$age), x$slaughter$args))
  pig_new$slaughtered <- is_slaughtered

  ## STEP 11: ageing of pigs not slaughtered in current cycle
  pig_new$age <- pig_new$age + 1

  ## STEP 12: waning of immunity of pigs not slaughtered in current cycle
  is_immune <- pig_new$immunity > 0
  pig_new$immunity[is_immune] <-
    pig_new$immunity[is_immune] - 1

  ## STEP 13: update 'time_since_vaccination' counter for vaccinated pigs
  is_vaccinated <- !is.na(pig_new$time_since_vaccination)
  pig_new$time_since_vaccination[is_vaccinated] <-
    pig_new$time_since_vaccination[is_vaccinated] - 1

  ## STEP 14: birth of new pigs
  ## .. only if pigs were slaughered
  if (sum(is_slaughtered) > 0) {
    pig_new <-
      rbind(pig_new,
            data.frame(
			  stringsAsFactors = FALSE,
			  age = rep(0, sum(is_slaughtered)),
              cysti = 0,
              cysti_immature = 0,
              time_since_infection = 0,
              intensity = 0,
              immunity = 0,
              time_since_vaccination = NA,
              slaughtered = 0))
  }

  ## replace data frames with new ones
  x$pig <- pig_new
  x$man <- man_new

  ### take populaton census
  out[i, ] <- census(x)

  ## update progress bar
  if (verbose) setTxtProgressBar(pb, i)
}

## close progress bar
if (verbose) close(pb)

## combine old and new census
x$out <- rbind(x$out, out)

## return updated 'cystiRun' object
return(x)
}


R/methods.R
### CYSTIRUN METHODS

print.cystiRun <-
function(x, from = 200, to = NA) {
  if (is.na(to)) to <- nrow(x$out)
  p <- colMeans(x$out[seq(from, to), ])
  cat("PC: ", round(p[1], 3),
      "| Pr(H):", round(p[2], 3), "\n")
  cat("PCI:", round(p[4], 3), "\n")
  cat("HT: ", round(p[6], 4),
      "| HT(inf):", round(p[7], 4),
      ", HT(chl):", round(p[8], 4),
      ", HT(adl):", round(p[9], 4), "\n")
  cat("E:  ", round(p[10], 4), "\n")
}

plot.cystiRun <-
function(x, show = c("PC", "PR", "HT", "EN"),
         start = 0, from = 1, to = NA) {
  ## define census labels and groups
  label <- c("PC", "P(H)", "P(L)", "PCI", "PR",
             "HT", "HT(inf)", "HT(ch)", "HT(ad)", "EN")
  group <- c(rep("pig", 5), rep("man", 5))

  ## select labels and groups to show
  id_col <- match(show, label)
  lab <- factor(label[id_col], levels = label[id_col])
  grp <- factor(group[id_col], levels = unique(group[id_col]))

  ## define extraction points
  if (is.na(to)) to <- nrow(x$out)
  id_row <- seq(from, to)

  ## extract census data; force matrix
  out <- as.matrix(x$out[id_row, id_col])

  ## define months
  months <- seq(nrow(out)) - start

  ## create ggplot data.frame
  df <- data.frame(p = c(out),
                   m = rep(months, times = ncol(out)),
                   lab = rep(lab, each = nrow(out)),
                   grp = rep(grp, each = nrow(out)))

  ## build plot function
  ggplot(df, aes(x = m, y = p)) +
    geom_line(aes(col = lab)) +
    scale_colour_manual(values = seq(10)[id_col]) +
    facet_grid(grp ~ ., scales = "free", as.table = F) +
    scale_x_continuous("month") +
    scale_y_continuous("proportion") +
    theme_bw()
}


R/baseline.R
### CYSTISIM: AGENT BASED TAENIA SOLIUM TRANSMISSION MODEL

## -------------------------------------------------------------------------#
## GENERATE RANDOM BASELINE ------------------------------------------------#

## simple random baseline for humans
random_baseline_man <-
function(n, p) {
  data.frame(
    stringsAsFactors = FALSE,
    age = round(runif(n, 0, 80) * 12),
    sex = sample(c("male", "female"), n, replace = TRUE),
    taenia = rbinom(n, 1, p),
    taenia_immature = 0,
    time_since_infection = 0,
    environment = 0,
    time_since_contamination = 0)
}

## slaughter function - default
slaughter_nbinom <-
function(age, min, max, size, mu) {
  ## if age below 'min', do not kill
  kill <- rep(0, length(age))

  ## if age larger or equal 'max', always kill
  kill[age >= max] <- 1

  ## if age in between, kill probabilistically
  kill_age <- age >= min & age < max
  kill[kill_age] <-
    rbinom(sum(kill_age), 1, pnbinom(age[kill_age] - min, size, mu = mu))

  ## return results
  return(kill)
}

## slaughter function - simplified
slaughter_binom <-
function(age, min, max, p) {
  ## if age below 'min', do not kill
  kill <- rep(0, length(age))

  ## if age larger or equal 'max', always kill
  kill[age >= max] <- 1

  ## if age in between, kill probabilistically
  kill_age <- age >= min & age < max
  kill[kill_age] <- rbinom(sum(kill_age), 1, p)

  ## return results
  return(kill)
}

## model pig age structure - default
pig_age_model <-
function(n, steps, size, mu) {
  ## create population with 'n' births per month
  pigs <- data.frame(age = rep(0, n))

  for (i in seq(5)) {
    ## ageing of pigs
    pigs$age <- pigs$age + 1

    ## new piglets get born
    pigs <- rbind(pigs, data.frame(age = rep(0, n)))
  }

  ## define progress bar
  pb <- txtProgressBar(max = steps, style = 3)

  ## run through cycles
  for (i in seq(steps)) {
    ## ageing of pigs
    pigs$age <- pigs$age + 1

    ## slaughter of pigs - default
    kill <- slaughter_nbinom(pigs$age, min = 6, max = 36, size, mu)

    ## remove slaughtered pigs from population
    pigs <- subset(pigs, !kill)

    ## new piglets get born
    ## number of births equal to number of killed pigs
    pigs <- rbind(pigs, data.frame(age = rep(0, sum(kill))))

    ## update progress bar
    setTxtProgressBar(pb, i)
  }

  ## close progress bar
  close(pb)

  ## return pigs dataframe
  return(pigs)
}

## random baseline pigs
random_baseline_pig <-
function(n, p, p.high, size = 0.70, mu = 80) {
  ## model age structure of pig population
  pigs_age <- pig_age_model(n / 6, 500, size, mu)$age

  ## identify pigs that are 3 months or older
  ## only these pigs are old enough to appear Ag positive
  pigs_inf <- pigs_age > 3

  ## randomly infect pigs
  cysti <- rep(0, length(pigs_inf))
  cysti[pigs_inf] <- rbinom(sum(pigs_inf), 1, p)

  ## randomly assign infection intensity to infected pigs
  intensity <- rep(0, length(pigs_inf))
  intensity[cysti == 1] <-
    sample(c("H", "L"), sum(cysti), TRUE, c(p.high, 1 - p.high))

  ## create 'pigs' dataframe
  pigs <-
    data.frame(
	  stringsAsFactors = FALSE,
	  age = pigs_age,
      cysti = cysti,
      cysti_immature = 0,
      time_since_infection = 0,
      intensity = intensity,
      immunity = 0,
      time_since_vaccination = NA,
      slaughtered = 0)

  ## return 'pigs' dataframe
  return(pigs)
}

R/cystiSim.R
### CYSTISIM: AGENT BASED TAENIA SOLIUM TRANSMISSION MODEL

## -------------------------------------------------------------------------#
## MAIN SIMULATION FUNCTION ------------------------------------------------#

cystiSim <-
function(n = 100, mod, main = NULL) {
  ## check if global variables are present
  ph2m <- get("ph2m", envir = .GlobalEnv)
  pl2m <- get("pl2m", envir = .GlobalEnv)
  m2p  <- get("m2p",  envir = .GlobalEnv)
  e2p  <- get("e2p", envir = .GlobalEnv)
  cov_man_mda <- get("cov_man_mda", envir = .GlobalEnv)
  cov_pig_mda <- get("cov_pig_mda", envir = .GlobalEnv)
  cov_pig_vac <- get("cov_pig_vac", envir = .GlobalEnv)
  eff_man_mda <- get("eff_man_mda", envir = .GlobalEnv)
  eff_pig_mda <- get("eff_pig_mda", envir = .GlobalEnv)
  eff_pig_vac <- get("eff_pig_vac", envir = .GlobalEnv)

  ## predefine list of simulations
  out <- vector("list", n)

  ## perform simulations
  for (i in seq(n)) {
    print(i)
    out[[i]] <- eval(substitute(mod))$out[, c(1, 5, 6)]
  }

  ## create 'cystiSim' object
  sim <-
    list(out = out,
         mod = substitute(mod),
         main = main,
         par = list(ph2m = ph2m,
                    pl2m = pl2m,
                    m2p = m2p,
                    e2p = e2p,
                    cov_man_mda = cov_man_mda,
                    cov_pig_mda = cov_pig_mda,
                    cov_pig_vac = cov_pig_vac,
                    eff_man_mda = eff_man_mda,
                    eff_pig_mda = eff_pig_mda,
                    eff_pig_vac = eff_pig_vac))

  ## add S3 class 'cystiSim'
  class(sim) <- "cystiSim"

  ## return 'cystiSim' object
  return(sim)
}


## -------------------------------------------------------------------------#
## cystiSim METHODS -------------------------------------------------------#

print.cystiSim <-
function(x, ...) {
  mod <- strsplit(as.character(x$mod)[2], "%>%")[[1]]
  mod <- gsub("\n", "", mod)
  mod <- sub("^ +", "", mod)
  mod <- sub(" +$", "", mod)
  for (i in seq_along(mod)) {
    if (i > 1) cat ("  ")
    cat(mod[i])
    if (i < length(mod)) cat(" %>%\n")
  }
  cat("\n")
}

summary.cystiSim <-
function(object, round = 3, ...) {
  sim_mx <- sapply(object$out, as.matrix)
  m <- nrow(sim_mx) / ncol(object$out[[1]])

  out <-
  rbind(
    c(mean(sim_mx[m, ] == 0), quantile(sim_mx[m, ], 0.95)),      # PC
    c(mean(sim_mx[3*m, ] == 0), quantile(sim_mx[3*m, ], 0.95)))  # HT

  rownames(out) <- c("PC", "HT")
  colnames(out) <- c("Pr(elim)", "95%")

  return(round(out, round))
}

plot.cystiSim <-
function(x, y = NULL, annotate = TRUE, ...) {
  lab <- factor(c("PC", "PR", "HT"), c("PC", "PR", "HT"))
  grp <- factor(c("pig", "pig", "human"), c("pig", "human"))
  col <- c(1, 3, 2)

  sim_mx <- sapply(x$out, as.matrix)  # ncol=it; nrow=par*months

  df <-
  data.frame(mean = rowMeans(sim_mx),
             lwr  = apply(sim_mx, 1, quantile, 0.025),
             upr  = apply(sim_mx, 1, quantile, 0.975),
             m = rep(seq(nrow(sim_mx) / ncol(x$out[[1]])),
                     ncol(x$out[[1]])),
             par = rep(lab, each = nrow(sim_mx) / ncol(x$out[[1]])),
             grp = rep(grp, each = nrow(sim_mx) / ncol(x$out[[1]])))

  stats <- summary(x)

  x_txt <- nrow(sim_mx) / ncol(x$out[[1]])
  y_txt <- tapply(apply(sim_mx, 1, max), df$grp, max)

  df_txt <-
    data.frame(lab = paste0(ncol(sim_mx), " simulations",
                            "\nPr(elim)=", stats[, 1]),
               grp = levels(df$grp),
               x = x_txt,
               y = y_txt)

  g <-
  ggplot(df, aes_string(x = "m", y = "mean")) +
    geom_ribbon(aes_string(ymin = "lwr", ymax = "upr", fill = "par"),
                alpha = .25) +
    geom_line(aes_string(y = "mean", col = "par"), size = 1) +
    geom_line(aes_string(y = "lwr", col = "par")) +
    geom_line(aes_string(y = "upr", col = "par")) +
    scale_fill_manual(values = col) +
    scale_colour_manual(values = col) +
    scale_x_continuous("month") +
    scale_y_continuous("prevalence") +
    facet_grid(grp~., scales = "free") +
    theme(legend.position = "none") +
    theme_bw() +
    ggtitle(x$main)

  if (annotate) {
    g <-
      g +
      geom_text(data = df_txt,
                aes_string(label = "lab", x = "x", y = "y"),
                size = 4, hjust = 1, vjust = 1)
  }

  return(g)
}

report <-
function(x, ...) {
  UseMethod("report")
}

report.cystiSim <-
function(x, name = "cystiSim", ...) {
  ## temporary 'mod' file
  ## write model
  tmp <- tempfile(fileext = ".Rnw")
  sink(tmp)
  cat("<<eval=FALSE>>=\n")  
  print(x)
  cat("@\n")
  sink()

  ## knit PDF
  knit2pdf(system.file("cystiSim.Rnw", package = "cystiSim"))

  ## rename PDF
  file.rename("cystiSim.pdf",
              paste0(name, "_", today(), ".pdf"))

  ## remove helper files
  unlink(tmp)
  file.remove(paste0("cystiSim.",
                     c("tex", "log", "aux")))
  unlink("figure", recursive = TRUE)

  ## generate PNG
  png(paste0(name, "_", today(), ".png"),
      width = 10, height = 5, units = "in", res = 300)
  print(plot(x))
  graphics.off()
}

elim <-
function(x, ...) {
  UseMethod("elim")
}

elim.cystiSim <-
function(x, show = c("m", "y"), ...) {
  show <- match.arg(show)
  cat(nrow(x$out[[1]]), "iterations\n")
  pc <- sapply(x$out, function(x) tail(which(x[, 1] != 0), 1)) - 201
  ht <- sapply(x$out, function(x) tail(which(x[, 3] != 0), 1)) - 201

  div <- ifelse(show == "m", 1, 12)

  Summary <-
  function(x) {
    c(mean = mean(x),
      min = min(x),
      max = max(x),
      quantile(x, c(0.025, 0.975)))
  }

  rbind(pc = Summary(pc), ht = Summary(ht)) / div
}

## -------------------------------------------------------------------------#
## HELPER FUNCTIONS --------------------------------------------------------#

today <-
function() {
  return(format(Sys.time(), "%Y%m%d"))
}